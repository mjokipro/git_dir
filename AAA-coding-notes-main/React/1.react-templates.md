
---
date: 2023-04-19
metadata: true
concepts: ['react', 'template']
status: 'pre-lecture'
docs: 
cite: ['rithm']
---


## App.js

```jsx
import React from "react";
import "./App.css";
import Component from "./Component";

function App() {
	return (
		<div className="App">
			<Snowman />
		</div>
	);
}

export default App;
```

## Component.js

```jsx
import { useState } from "react";
import "./Component.css";
import Subcomponent from "./Subcomponent";
import { v4 as uuid } from "uuid";

/** Order entering system before it ships.
 *
 * Props:
 * - orderId
 * - price (before tax)
 * - salespersonId
 *
 * State:
 * - isConfirmed: true/false
 *
 * Customer -> Order -> OrderItem
 */
function Component({props}){
  const [playerHand, setPlayerHand] = useState(getPair);
  const [dealerHand, setDealerHand] = useState([]);
  
    /** Player draws a card and then dealer plays. */
  function takeCard() {
    setPlayerHand(h => [...h, getRandomCard()]);
    playDealer();
  }
  
  /** Dealer draws a pair and then hand is scored. */
  function playDealer() {
    setDealerHand(getPair());
  }

 const result = dealerHand.length > 0 ? score(playerHand, dealerHand) : null;
  
  return (
      <main>
        <Hand cards={dealerHand} />
        <Hand cards={playerHand} />
        {result === null && <p>
          <button id="draw" onClick={takeCard}>Draw</button>
          <button id="pass" onClick={playDealer}>Pass</button>
        </p>}
        {result !== null && <p>{result}</p>}
      </main>
  );
}
```

Form components:
```jsx
/** Update form input. */
function handleChange(evt) {
	const input = evt.target;
	setFormData(formData => ({
		...formData,
		[input.name]: input.value,
	}));
}

  

/** Call parent function and clear form. */
function handleSubmit(evt) {
	evt.preventDefault();
	handleSave(formData);
	setFormData(initialFormData);
}
```

## Component.test.js

- The render method returns many things, of which weâ€™ll use two:
	-   **container**: an DOM component of a div that holds what you rendered
	-   **debug(elem)**: a method that returns debugging information about elem
	-  [Full list](https://testing-library.com/docs/react-testing-library/api/#render-result)

```jsx
import { render, fireEvent } from "@testing-library/react";
import Component from "./Component";
import TEST_IMAGES from "./_testCommon.js"; // any other components/imgs

// smoke test
it("renders without crashing", function () {
  // this is a low-value test, but better than nothing
  render(<Card rank="A" suit="C" />);
});

// snapshot test - 
it("matches snapshot", function () {
  const { container } = render(<Card rank="A" suit="C" />);
  expect(container).toMatchSnapshot();
});

// snapshot test with mocking
  it("matches initial player hand: 2C 3C", function () {
    _feedChoice("2C 3C"); // Just abstracts away mocking fx code
    const { container } = render(<NineteenGame />);
    expect(container).toMatchSnapshot();
  });

// testing the DOM
it("has the correct alt text & src", function () {
  const { container, debug } = render(<Card rank="A" suit="C" />);
  const img = container.querySelector("img");
  debug(img);

  expect(img.getAttribute("alt")).toEqual("AC");
  expect(img.getAttribute("src")).toContain("AC.png");
});

// testing interactions
it("works when you click on the left/right arrows", function () {
	const { container } = render(
		<Carousel photos={TEST_IMAGES} title="images for testing" />
	);
	
	// expect first image to show, second image to not
	expect(
		container.querySelector('img[alt="testing image 1"]')
	).toBeInTheDocument();
		expect(
		container.querySelector('img[alt="testing image 2"]')
	).not.toBeInTheDocument();

	// click right arrow
	const rightArrow = container.querySelector(".bi-arrow-right-circle");
	fireEvent.click(rightArrow);
	
	// expect first image to be gone, and second image to show
	expect(
		container.querySelector('img[alt="testing image 1"]')
	).not.toBeInTheDocument();
	expect(
		container.querySelector('img[alt="testing image 2"]')
	).toBeInTheDocument();

	// click left arrow
	const leftArrow = container.querySelector(".bi-arrow-left-circle");
	fireEvent.click(leftArrow);

	// expect first image to show, second image to not
	expect(
		container.querySelector('img[alt="testing image 1"]')
	).toBeInTheDocument();
		expect(
		container.querySelector('img[alt="testing image 2"]')
	).not.toBeInTheDocument();
});

// tests with mocking
it("tests that mock function has been called", function () {
  const mocksave = jest.fn();
  render(<TodoForm handleSave={mockSave} />);

  expect(mockSave).toHaveBeenCalled();
});

it("stores priority as an integer", function () {
  // Other option - you hand write your own fx to mock to check input data
  function handleSave(formData) {
	expect(formData.priority).toEqual(expect.any(Number));
  }
  
  const { container } = render(<TodoForm handleSave={handleSave} />);

  const submitFormButton = container.querySelector("TodoForm-addBtn")
  fireEvent.click(submitFormButton);
  
  expect.assertions(1);
```

