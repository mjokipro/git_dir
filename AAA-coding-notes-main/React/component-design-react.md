
---
date: 2023-04-19
metadata: true
concepts: []
status: 'pre-lecture'
docs: 
cite: ['rithm']
---

## Goals

-   Understand strategies for ‚Äúcomponent decomposition‚Äù
-   Practice designing a React app!
-   Learn to how pass functions between components

## Component hierarchies

### Lucky 7 Game

demo/dice-game/src/Lucky7.js
```jsx
import { useState } from "react";
import { getRolls, sum } from "./utils";
import "./Lucky7.css";

/** Luck Game: roll 2d6 and win with 7! */

function Lucky7() {
  const [dice, setDice] = useState(getRolls(2));
  const won = sum(dice) === 7;

  function roll() { setDice(getRolls(2)); }

  return (
    <main className="Lucky7">
      <h1>Lucky7 {won && "You won!"}</h1>
      <section className="Lucky7-dice">
        <div className="Lucky7-die">{dice[0]}</div>
        <div className="Lucky7-die">{dice[1]}</div>
      </section>
      <button onClick={roll}>
        Roll Again!
      </button>
    </main>
  );
}
```

-   This component is doing a lot!
-   It‚Äôs also inflexible
    -   It always rolls 2 dice
    -   It always wins on 7
-   There aren‚Äôt reusable parts
    -   Showing a set of dice
    -   Show an individual die
-   Let‚Äôs do better!

### Requirements

- A set components for:
	-   Play a dice game with _numDice_ number of dice
	-   Shows a win message when total equals _goal_
	-   A ‚Äúroll again‚Äù button that re-rolls all the dice
- What props do we need? What state do we need? And where?

![](../assets/image/component-design-react-1681938340429.jpeg)

### Die Component

-   Props
    -   val: number
-   State
    -   none!
-   Events
    -   none!

![](../assets/image/component-design-react-1681938374341.jpeg)

die component
```jsx
/** Single die. */

function Die({ val }) {
  return (
      <div className="Die">
        {val}
      </div>
  );
}
```

### Dice Component

-   Props
    -   dice: `[n, ...]`
-   State
    -   none!
-   Events
    -   none!

![](../assets/image/component-design-react-1681938415133.jpeg)

dice component
```jsx
/** Shows a set of dice. */

function Dice({ dice }) {
  return (
    <section className="Dice">
      {dice.map((v, i) =>
        <Die key={i} val={v} />)}
    </section>
  );
}
```

### LuckyN Component

-   Props
    -   numDice
    -   goal
-   State
    -   dice: `[n, ...]`
-   Events
    -   roll()

luckyn
```jsx
/** Luck Game: roll numDice & d6 and win with goal! */

function LuckyN({ numDice, goal }) {
  const [dice, setDice] = useState(getRolls(numDice));
  const won = sum(dice) === goal;

  function roll() { setDice(getRolls(numDice)); }

  return (
      <main className="LuckyN">
        <h1>Lucky{goal} {won && "You won!"}</h1>
        <Dice dice={dice} />
        <button onClick={roll}>
          Roll Again!
        </button>
      </main>
  );
}

```

### App Component

-   Props
    -   none!
-   State
    -   none!
-   Events
    -   none!

![](../assets/image/component-design-react-1681938503042.jpeg)

demo/dice-game/src/App.js
```jsx
/** Site with dice games. */

function App() {
  return (
    <div className="App">
      <Lucky7 />
    </div>
  );
}
```

### State design principle

> Lift state as high as needed ‚Äî  
> but no higher.

Where should the dice-rolls state live?

- [‚ùå|25](https://twemoji.maxcdn.com/v/14.0.2/svg/274c.svg)App: don‚Äôt need it, so shouldn‚Äôt lift it
-   ![‚úÖ|25](https://twemoji.maxcdn.com/v/14.0.2/svg/2705.svg) LuckyN: this is the game itself!
-   ![‚ùå|25](https://twemoji.maxcdn.com/v/14.0.2/svg/274c.svg) Dice: should just be about showing a hand
-   ![‚ùå]|25(https://twemoji.maxcdn.com/v/14.0.2/svg/274c.svg) Die: need to know roll total; not just for one>)

- & lift the state component up as high as needed - but no higher.

![](../assets/image/component-design-react-1681938729776.jpeg)

### Decoupling logic from presentation

- Generally, aim to have components be one of two types:
	- Presentational ![üé®|25](https://twemoji.maxcdn.com/v/14.0.2/svg/1f3a8.svg)
		- Doesn‚Äôt have state; is primarily about appearance/UI.
		- ‚Äúdumb components‚Äù
	- Logical ![üß†|25](https://twemoji.maxcdn.com/v/14.0.2/svg/1f9e0.svg)
		- Has state or related logic; isn‚Äôt about a specific UI.

![](../assets/image/component-design-react-1681938715169.jpeg)

## Getting logic out of views

demo/dice-game/src/utils.js
```jsx
/** Gets random integer: [1..6]. */

function d6() {
  return Math.floor(Math.random() * 6) + 1;
}

/** Get n rolls => [num, ...]. */

function getRolls(n) {
  return Array.from({length: n}, () => d6());
}

/** Get sum of nums. */

function sum(nums) {
  return nums.reduce((prev, cur) => prev + cur, 0);
}

export { d6, getRolls, sum };
```

- Not everything needs to done in your component function:
	-   common logic can be shared
	-   even if only used once,  
	    this can still be very useful:
	    -   can be unit tested
	    -   can be mocked

### LuckyN

```jsx
/** Luck Game: roll numDice & d6 and win with goal! */

function LuckyN({ numDice, goal }) {
  const [dice, setDice] = useState(getRolls(numDice));
  const won = sum(dice) === goal;

  function roll() { setDice(getRolls(numDice)); }

  return (
      <main className="LuckyN">
        <h1>Lucky{goal} {won && "You won!"}</h1>
        <Dice dice={dice} />
        <button onClick={roll}>
          Roll Again!
        </button>
      </main>
  );
}
```

-   LuckyN takes numDice and goal
-   What if we wanted a more different game?
-   One where you had to roll less than a number?
-   Right now, logic for winning is in this function

### Passing a function as a prop

- In JavaScript, functions are _first-class objects_ ‚Äî you can pass them around!
- LuckyN Component is about weby stuff - how it renders and how it looks - and now it doesn‚Äôt actually need to know how the game is won - this is now contained in the function 
- 

- We can pass a winCheck(dice) function to our LuckyN game!
```jsx
/** Possible game strategy: sum of dice < 4. */
function lessThan4(dice) {
  return sum(dice) < 4;
}

<LuckyN numDice={3} winCheck={lessThan4} />
```

```jsx
/** Luck Game: roll numDice & d6 and win with goal! */

function LuckyN({ numDice, winCheck }) {
  const [dice, setDice] = useState(getRolls(numDice));
  const won = winCheck(dice);

  // ...
}
```

- ~ Note: That could also be inlined
	- Given how simple the logic for that particular win condition is, this could even be inlined in the rendering of the function:
	```jsx
	<LuckyN
	  numDice={3}
	  winCheck={dice => sum(dice) < 4)} />
	```
	- However, that would make it much harder to unit test the condition or mock it, so this is probably not a good idea.

- & This is a very powerful idea ‚Äî now, our LuckyN component is about the  _view_ our our app, and it can take *configurable logic*.

## Decomposing components

![](../assets/image/component-design-react-1681939476562.jpeg)

-   Let‚Äôs make this button fancier
	- But LuckyN about logic/state, not UI
-   Also: we might like to re-use this button elsewhere
-   Let‚Äôs move it to its own component!
	- Lets you use fancy button for whatever thing you want it to do -
		- That‚Äôs the fx you pass to it

### Passing functions to child

-   It‚Äôs common to pass a function to a child, for them to execute as needed.
-   Our dice state is in the game, so button can‚Äôt change it directly.
-   Our roll() function changes the state, so game with re-render.

### Making button more reusable

- Even better ‚Äî we can make our button even more generic!
	- Click - function called onClick
	- Label - the label on button
![](../assets/image/component-design-react-1681939559497.jpeg)

## Function passing

### New feature: re-rolling individual dice

![](../assets/image/component-design-react-1681940010714.jpeg)

- How could we add a feature to re-roll an individual die when it‚Äôs clicked on?
- We can‚Äôt add it to LuckyNButton2 _or_ Die without changes!

![](../assets/image/component-design-react-1681940028473.jpeg)
-   Add rollDie(n) to game component
-   Pass it to Die _through_ Dice
-   Each Die needs to know its ‚Äúposition‚Äù from Dice loop
-   On click, Die will call roll() which calls rollDie(myIndex)

![](../assets/image/component-design-react-1681940078836.jpeg)

### How data flows

- This shows a common pattern in React:
	-   A parent component defines a function
	-   The function is passed as a prop to a child component
	-   The child component invokes the prop function
	-   The parent function is called, usually setting new state
	-   The parent component is re-rendered along with its children

- ~ Alternative strategy: pass ‚Äúcustom function‚Äù to Die
	- Right now, DiceRR passes rollDie() through Dice to each Die.  
	- Dice has to tell each die their index:
	LuckyRR
	```jsx
	<DiceRR dice={dice} rollDie={rollDie}/>
	```
	DiceRR
	```jsx
	{dice.map((v, idx) =>
	  <DieRR idx={idx} key={idx} rollDie={rollDie} val={v} />
	)}
	```
	DieRR
	```jsx
	function DieRR({ idx, rollDie, val }) {
	  function roll() { rollDie(idx); }
	
	  return <div className="Die" onClick={roll}>{val}</div>;
	}
	```
	- We could instead pass a _custom function_ to each child:
	LuckyRR
	```jsx
	<DiceRR dice={dice} rollDie={rollDie}/>  // unchanged
	```
	DiceRR
	```jsx
	{dice.map((v, idx) =>
	  // making bespoke `rollMe` for each Die
	  <DieRR idx={idx} rollMe={idx => rollDie(idx)} val={v} />
	)}
	```
	DieRR
	```jsx
	function DieRR({ rollMe, val }) {
	  return <div className="Die" onClick={rollMe}>{val}</div>;
	}
	```
	- Which is better?
		-   Using arrow functions in the parent simplifies the child
		-   However, there are performance optimization considerations
		-   Prefer passing rollDie to Die over a custom arrow function for now

## Learning design

- Designing a React application is a new skill will take practice to master.

### Components

- Generally, components should be small & do one thing.
- This often makes them more reusable.
- Example: component that displays a todo w/task could be used in lots of ‚Äúlists‚Äù.

### Presentational Components

- Often, small components are simple & don‚Äôt have state:
```jsx
function Todo(props) {
  return <div className="Todo">{ props.task }</div>;
}
```

- This can be used like:
```jsx
function ListOfTodos() {  // ... lots missing
  return (
    <div className="ListOfTodos">
      todos.map(t => <Todo task={t} />
    </div>
  );
}
```

- Components like Todo are called ‚Äúpresentational‚Äù or ‚Äúdumb‚Äù _\[in a good way!\]_

### Don‚Äôt Store Derived Info

- & If one thing can be calculated from another, don‚Äôt store both:
```jsx
function TaskList() {
  const [todos, setTodos] = useState(["wash car", "wash cat"]);
  const [numTodos, setNumTodos] = useState(2);

  return (
    <div>
      You have {numTodos} tasks ...
    </div>
  );
}
```

- ! Yuck! Just calculate the number of todos as needed!

## Coming Up

-   Learning to ‚Äúroute‚Äù in React applications
-   Learning how to handle ‚Äúside effects‚Äù in a component
-   Learning other ways to pass data around

- ~ Using arrow functions for components
	- Components are just functions. We can write them with arrow syntax if we choose.
	If the component only renders, you can make use of an arrow function‚Äôs implicit return.
	![](../assets/image/component-design-react-1681940499667.jpeg)
	- *Should you use arrow functions for components?* We recommend ‚úó no ‚Äî but, as always, follow the style of existing code.
	- You should understand the syntax so you recognize it in some documentation, code examples, etc.
	- Whichever you choose: be consistent.



## Additional Component Design Examples:

- From *Productiv* mini-sprint: 
![](../assets/image/component-design-react-1682681112627.jpeg)
![](../assets/image/component-design-react-1682681202810.jpeg)
![](../assets/image/component-design-react-1682681228763.jpeg)
