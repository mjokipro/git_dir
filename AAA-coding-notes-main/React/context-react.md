
---
date: 2023-04-27
metadata: true
concepts: ['react', 'context']
status: 'pre-lecture'
docs: 
cite: ['rithm']
---

## Goals

-   Explain what context is
-   Use the Context API to provide and consume context

## Motivation

### What is Context?

-   Universal data across your application
-   Data accessible across all components

### When Is It Useful?

- Useful for global themes, shared data:
	-   Passing a user’s name around
	-   User currency preferences
	-   Light or dark mode theme selection

## Creating Context

demo/user-context/src/userContext.js
```jsx
import { createContext } from "react";

const userContext = createContext();

export default userContext;
```

- This gives us a component:
-   `<userContext.Provider>`: can provide a value to context

### Provider

- Note: context created inside of App.js
demo/user-context/src/JoblyApp.js
```jsx nums {5, 17, 21}
import { useState } from "react";
import UserPrefForm from "./UserPrefForm";
import JobDetail from "./JobDetail";
import CompanyDetail from "./CompanyDetail";
import userContext from "./userContext";

const DEFAULT_PREFS = {color: "dark", currency: "USD"};

function JoblyApp() {
    const [prefs, setPrefs] = useState(DEFAULT_PREFS);

    function updatePrefs(newPrefs) {
        setPrefs(newPrefs);
    }

    return (
    <userContext.Provider value={{user: null, prefs}}>
      <UserPrefForm submit={updatePrefs} currPrefs={prefs} />
      <JobDetail />
      <CompanyDetail />
    </userContext.Provider>
  );
}
```

- Any component descending from the Provider can access the context value.
	- & Meaning any component that has the `<userContext.Provider etc...>` wrapped around it *can access that context*

### useContext

demo/user-context/src/JobDetail.js
```jsx nums {1-2, 5, 10}
import { useContext } from "react";
import userContext from "./userContext";

function JobDetail({ title, description, salary }) {
  const { prefs } = useContext(userContext);

  return (
    <div>
      <h2>Fake Job!</h2>
      <p>You should apply! It pays lots of {prefs.currency}.</p>
    </div>
  )
}
```

- & In order to get the value, we need the `useContext `hook.
-   `useContext` looks for a matching context, and reads its value.
-   When value inside of context changes, components accessing that context will re-render.
-   Components that read from context with `useContext` are sometimes called _consumers_ (as opposed to providers).

### Storing User in Context

```jsx nums {15}
function JoblyApp() {
  // same as before but also...

  const [user, setUser] = useState(null);

  function login({username, password}) {
    // lots of logic here, but then ...
    const userFromAPI = {get_this_from_api};
    setUser(userFromAPI);
  }

  return (
  <userContext.Provider value={{user, prefs}}>
    {/* same, plus ... */}
    <LoginForm submit={login} />
  </userContext.Provider>
  );
}
```

- ! Be careful here about what information gets stored in context
	- $ username, name –> probably fine
	- !! SSN, mother’s maiden name, etc. 

- & Usually context is a piece of state
	- If you want to update that state - it needs to be updated in the same place that the context is defined.

## Demo Time

### Deadly Doubles

- A casino of different dice table games.

- Try it out: [https://react-deadly-doubles.surge.sh/](https://react-deadly-doubles.surge.sh/)

### Without Using Context

![](../assets/image/context-react-1682597330058.jpeg)
- Can you see the challenge?
- All components between Casino and Die don’t care about color for dice

### Using Context

![](../assets/image/context-react-1682597456542.jpeg)

- Now those components don’t need to pass down props they don’t care about

### React Features

- A generic component, *Table*, which can render different games.
- React’s context manager: *Casino* lets you choose a favorite color,  and *Die* (several layers down) can access the color you chose.

### Should I Use Context For Passing Everything?

- No!
- Passing the “real data” down as traditional props is better — it’s clear, obvious, and how React is meant to be used.
	- Note, like in earlier example, if something is used explicitly, should be passed explicity as a prop anyway
	- Here, prefs is pass to UserPrefForm explicity even though it is available in context:
![](../assets/image/context-react-1682616432488.jpeg)
- Context is good for cases where lots of different components need incidental access to *non-core-data*, and is convenient for them to to it easily.
	- Maybe don’t put the whole user object in context, but just username or aspects that are used and needed. 


## Documentation for context

- Put in docstring if context is relied on for that particular component 
	- But no need to document this for the in-between components
![](../assets/image/context-react-1682597456542.jpeg)
- $ So in this example, would document it for:
	- $ Casino - where context stored
	- $ Die - that relies on the color stored in context
- !! But Not:
	- !! Tables, Table , DeadlyDoubles, DiceSet
