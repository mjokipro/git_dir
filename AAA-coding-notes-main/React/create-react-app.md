
---
date: 2023-04-17
metadata: true
concepts: ['react', 'create-react-app']
status: 'pre-lecture'
docs: 
cite: ['rithm']
---
 
## Goals

-   Understand what **Create-React-App** is and how to use it
-   Use ES2015 modules to share code across files
-   Compare default vs. non-default exports
-   Using assets (images and CSS) in components

## Create React App

- React is a front-end library — you don’t need server-side stuff.
- You _can_ get `react.js` and `react-dom.js` from a CDN.
- You _can_ transpile JSX in the browser at runtime.
- But there’s a better way!

- **Create-React-App** is a utility script that:
	-   Creates a skeleton React project
	-   Sets it up so that JS files are run through Babel automatically
	-   Lets us use super-modern JavaScript features/idioms
	-   Makes testing & deployment much easier

### npx

- To scaffold a project with Create React App, we’ll use **npx**.
- **npx** will download Create React App and execute it.
- You can think of **npx** as being an alternative to installing packages globally.

Example
```shell
$ npx create-react-app my-app-name
```

### Skeleton

- This provides a nice starter skeleton:
```
├── README.md               README, can edit or delete
├── package-lock.json       Lock file, don't edit directly
├── package.json            Can edit, as usual
│
├── public                  Rarely need to edit these
│   ├── favicon.ico
│   ├── index.html          Main HTML page of site
│   ├── logo192.png
│   ├── logo512.png
│   ├── manifest.json
│   └── robots.txt
│
└── src                     Where React stuff goes
    ├── App.css                 CSS for example component
    ├── App.js                  Example component
    ├── App.test.js             Tests for App component
    ├── index.css               Site-wide CSS
    ├── index.js                Start of JS
    ├── logo.svg                React logo
    ├── reportWebVitals.js      (Ignore this for now)
    └── setupTests.js           Starter test configuration
```

![](../assets/image/create-react-app-1681763363723.jpeg)

### Starting Your App

```shell
$ npm start
```

- & Important: Remove Boilerplate Code
	- CRA will create a basic, boilerplate app for you – you should go ahead and remove the things you don’t need:
		- & Change the App component so it returns an empty div.
		- Delete the logo.svg file.
		- Delete the favicon.ico file (or replace with your own!)
		- & Remove the <React.StrictMode> tags around the <App /> component in index.js. React Strict mode will render _all of your components twice_, to test that they will render the same way when given them same input. This can be confusing to beginners who are still learning how React works, and when each component is rendered.

- create-react-app does strict mode for you

### Webpack

- CRA is built on top of Webpack, a JS utility that:
	- Enables module importing/exporting
	    - Packages up all CSS/images/JS into a single file for browser
	    - Dramatically reduces # of HTTP requests for performance
	- Hot reloading: when you change a source file, automatically reloads
	    - Is very clever and tries to only reload relevant files
	- Enables easy testing & deployment

- ~ Note: *The Webpack Rabbit Hole*
	- Webpack is a powerful tool, and configuring it can be quite complicated. Create React App abstracts away that configuration from you, which is great when you’re first learning. It’s not worth your time right now to learn too much about webpack other than the high-level bullet points we’ve outlined. If you’re curious, you can always to go to the [Webpack website](https://webpack.js.org/), but be warned: Webpack is a rabbit hole it’s easy to go down and isn’t terribly important at this stage in your learning.

## Modules

- CRA uses ES2015 “modules”
- This is a newer, standardized version of Node’s `require()`
- You use this to export/import classes/data/functions between JS files

### Sample Component

demo/my-app-name/src/App.js
```jsx nums {1-3, 22}
import React from 'react';
import logo from './logo.svg';
import './App.css';

function App() {
  return (
    <div className="App">
      <header className="App-header">
        <img src={logo} className="App-logo" alt="logo" />
        <p>Edit <code>src/App.js</code> and save to reload.</p>
        <a
          className="App-link"
          href="https://reactjs.org"
        >
          Learn React
        </a>
      </header>
    </div>
  );
}

export default App;
```

### Importing “Default” Export

- Export default is going to be the *main* thing you are exporting from that file. 
- You can only export *one* thing as the default, though you can export other things as well.
	- In this case, don’t need curly braces
	- ! When importing a default export - possible to rename, but really bad idea to do so. 

demo/import-export/mystuff.js
```jsx
function myFunc() {
  console.log("Hi");
}

export default myFunc;
```

demo/import-export/index.js
```jsx
// Must start with dot --- "mystuff" would be a npm module!

import myFunc from './mystuff';
```

### Importing Non-Default Named Things

demo/import-export/mythings.js
```jsx
function otherFunc() { 
  console.log("Hey");
}

const luckyNumber = 13;

export { otherFunc, luckyNumber };
```

demo/import-export/index.js
```jsx
import { otherFunc, luckyNumber} from "./mythings";
```

### Importing Both

demo/import-export/both.js
```jsx
function mainFunc() {
  console.log("Ok");
}

const msg = "Awesome!";

export default mainFunc;
export { msg };
```

demo/import-export/index.js
```jsx
import mainFunc, { msg } from "./both";
```

### To Default or Not?

- Joel’s opinion and Minority opinion:
	- if export default, it makes things a little more confusing
	- other opinion is that exporting default provides a way of declaring this is the main thing. 

- Conventionally, default exports are used  
    when there’s a “most likely” thing to export.
- For example, in a React component file, it’s common to have the component be the default export.
- You never *need* a default export, but it can be helpful to indicate most important thing in a file.

## CRA and Components

- Good style:
	- Each React component goes in separate file
	    - `src/Car.js` for Car component
	    - `src/House.js` for House component
	- Define your function component, then export it as the default
	- Skeleton assumes top object is App in `App.js`
	    - Best to keep this

## Assets and CRA

- To include images and CSS, you can import them in JS files!
demo/my-app-name/src/App.js
```jsx nums {2-3, 9}
import React from 'react';
import logo from './logo.svg';
import './App.css';

function App() {
  return (
    <div className="App">
      <header className="App-header">
        <img src={logo} className="App-logo" alt="logo" />
        <p>Edit <code>src/App.js</code> and save to reload.</p>
        <a
          className="App-link"
          href="https://reactjs.org"
        >
          Learn React
        </a>
      </header>
    </div>
  );
}

export default App;
```

### CSS

- Make a CSS file for each React component
    - `House.css` for House component
- Import it at the top of `House.js`
    - Create-React-App will automatically load that CSS
- & Conventional to add `className="House"` onto House div
    - And use that as prefix for sub-items to style:
```jsx
<div className="House">
  <p className="House-title">{ props.title }</p>
  <p className="House-address">{ props.addr }</p>
</div>
```

### Images

-   Store images in src/ folder with the components
-   Load them where needed and use imported name where path should go:

Animal.js
```jsx
import puppy from "./puppy.jpg";

function Animal() {
  return (
	<div>
	  <img src={puppy} alt="Cute puppy!" />
	</div>
  );
}
```

### Building for Deployment

- `npm run build` makes `build/` folder of static files
- You can serve from a web server.
