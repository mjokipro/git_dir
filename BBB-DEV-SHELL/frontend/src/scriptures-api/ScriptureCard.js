import React from "react";

const ScriptureCard = ({key, book_name, chapter, verse, text}) => {
    
    return (
        <div className="container">            <h1><span>{book_name} {chapter}: {verse}</span></h1>        
            <p className="lead">{text}</p>
        </div>
    )
}

export default ScriptureCard