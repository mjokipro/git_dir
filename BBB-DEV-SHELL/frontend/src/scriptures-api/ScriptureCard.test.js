
import React from "react";
import { render } from "@testing-library/react";
import ScriptureCard from './ScriptureCard'
import { MemoryRouter } from "react-router";

it("matches snapshot with logo", function () {
  const { asFragment } = render(
      <MemoryRouter>
        <ScriptureCard
            // to_user="test"
            // from_user="test"
            // body="test"
        />
      </MemoryRouter>,
  );
  expect(asFragment()).toMatchSnapshot();
});
