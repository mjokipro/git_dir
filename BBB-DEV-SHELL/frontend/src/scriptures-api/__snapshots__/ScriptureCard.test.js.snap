// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot with logo 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
                
    <h1>
      <span>
         : 
      </span>
    </h1>
    <p
      class="lead"
    />
  </div>
</DocumentFragment>
`;
